PRINCIPLES.md

```diff
@@ -0,0 +1,259 @@
+<!-- see https://github.com/yzhang-gh/vscode-markdown/blob/master/README.md#table-of-contents -->
+# GitOps Principles v0.1.0 <!-- omit in toc -->
+
+## ⚠️ THIS DOCUMENT IS A WORK IN PROGRESS AND SUBJECT TO PUBLIC REVIEW BEFORE PUBLICATION <!-- omit in toc --> ⚠️ 
+
+This document provides a concrete definition of GitOps principles.
+
+The GitOps principles governs how humans and technical systems should interact in order to achieve the desired operational outcomes of repeatability, auditability, and visibility.
+
+The GitOps Principles are vendor and implementation neutral, they aim to provide a common framework of understanding regarding software operations.
+
+## Table of Content <!-- omit in toc -->
+
+<!-- Insert auto-generated toc here -->
+- [Summary](#summary)
+- [Introduction](#introduction)
+- [Scope](#scope)
+- [The GitOps Principles](#the-gitops-principles)
+  - [1. Declarative configuration](#1-declarative-configuration)
+    - [What is a system's Desired State?](#what-is-a-systems-desired-state)
+    - [Why must the Desired State be declarative data?](#why-must-the-desired-state-be-declarative-data)
+    - [Why is human readability required?](#why-is-human-readability-required)
+    - [How much of a system must be declared?](#how-much-of-a-system-must-be-declared)
+  - [2. Immutable configuration versions](#2-immutable-configuration-versions)
+    - [What forms a version?](#what-forms-a-version)
+  - [3. Continuous state reconciliation](#3-continuous-state-reconciliation)
+  - [4. Operations through declaration](#4-operations-through-declaration)
+- [See Also](#see-also)
+
+## Summary
+
+GitOps is a set of principles for operating and managing software systems.
+
+When using GitOps, the desired state of a system or subsystem is defined declaratively as versioned, immutable data, and the running system's configuration is continuously derived from this data.
+
+GitOps principles were derived from modern software operations but are rooted in pre-existing and widely adopted best practices. These principles are:
+
+1. [**The principle of declarative configuration**](#1-declarative-configuration)
+
+    A system managed by GitOps must have its _Desired State_ expressed declaritively as data in a format writable and readable by both humans and machines.
+
+2. [**The principle of immutable configuration versions**](#2-immutable-configuration-versions)
+
+    _Desired State_ is stored in a way that supports versioning, immutability of versions, and retains a complete version history.
```

Comment:
```suggestion
    _Desired State_ is stored in a way that may support versioning and immutability of versions, and may retain a complete version history.
```
Suggested adding multiple "may"s because this principle shouldn't prevent practitioners from implementing Git, GitHub, BFG features such as:
1. [Configuring commit squashing for pull requests](https://docs.github.com/en/github/administering-a-repository/configuring-commit-squashing-for-pull-requests): `squash`
2. [Git Tools - Rewriting History](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History): `commit --amend`, `rebase`, `filter-branch`
3. [BFG is a simpler, faster alternative to git-filter-branch for cleansing bad data out of your Git repository history](https://rtyley.github.io/bfg-repo-cleaner/): `bfg`

**Why:**
• While versioning, immutability, completeness are recommended, practitioners should have the flexibility to opt-out.
• Seasoned `git` users are aware of situations that justify usages of `squash`, `commit --amend`, `rebase`, `filter-branch`, `bfg`.

This relates to https://github.com/gitops-working-group/gitops-working-group/discussions/93#discussioncomment-642197

Thank you @bricef @scottrigby @moshloop @o6uoq

Reply:
I don't think any form of rewriting history should be allowed in the desired state store.

Rewriting history is perfectly acceptable *before* it merges into master (the desired state) but once it is desired state you can't erase it.

Likewise, filter-branch and bfg should not be allowed, if you commit a secret then you add a new commit to remove it and then revoke the secret. 

If you commit a large file or binary, then you either live with it, or you go off the GitOps path, rewrite history, and then go back on the GitOps path.

Reply:
Can we agree to disagree?  Thank you @moshloop @roberthstrand @splushii

I believe rewriting history is appropriate to remove private information.

For example:
> **GitHub Private Information Removal Policy**
>
> We offer this private information removal process as an exceptional service only for high-risk content that violates [GitHub's Terms of Service](https://docs.github.com/en/github/site-policy/github-acceptable-use-policies#3-conduct-restrictions), such as when your security is at risk from exposed access credentials. This guide describes the information GitHub needs from you in order to process a request to remove private information from a repository.
>
> **[What is Private Information?](https://docs.github.com/en/github/site-policy/github-private-information-removal-policy#what-is-private-information)**
>
> For the purposes of this document, “private information” refers to content that (i) should have been kept confidential, _and_ (ii) whose public availability poses a specific or targeted security risk to you or your organization.
>
> "Security risk" refers to a situation involving exposure to physical danger, identity theft, or increased likelihood of unauthorized access to physical or network facilities.
>
> [Private information removal requests are appropriate for:](https://docs.github.com/en/github/site-policy/github-private-information-removal-policy#private-information-removal-requests-are-appropriate-for)
> - Access credentials, such as user names combined with passwords, access tokens, or other sensitive secrets that can grant access to your organization's server, network, or domain.
> - AWS tokens and other similar access credentials that grant access to a third party on your behalf. You must be able to show that the token does belong to you.
> - Documentation (such as network diagrams or architecture) that poses a specific security risk for an organization.
> • [Information](https://docs.github.com/en/github/site-policy/github-community-guidelines#doxxing-and-invasion-of-privacy) related to, and posing a security risk to, you as an individual (such as social security numbers or other government identification numbers).